cpp/src/arrow/type_test.cc

```diff
@@ -134,14 +189,40 @@ TEST(TestField, Equals) {
   AssertFieldsEqual(f0_with_meta1, f0_with_meta2, false);
 }
 
+TEST(TestField, IsCompatibleWith) {
+  auto meta1 = key_value_metadata({{"a", "1"}, {"b", "2"}});
+  // Different from meta1
+  auto meta2 = key_value_metadata({{"a", "1"}, {"b", "3"}});
+  // Equal to meta1, though in different order
+  auto meta3 = key_value_metadata({{"b", "2"}, {"a", "1"}});
+
+  Field f0("f0", int32());
+  Field f0_nn("f0", int32(), false);
+  Field f0_nt("f0", null());
+  Field f0_other("f0", int32());
+  Field f0_with_meta1("f0", int32(), true, meta1);
+  Field f0_with_meta2("f0", int32(), true, meta2);
+  Field f0_with_meta3("f0", int32(), true, meta3);
+  Field other("other", int64());
+
+  AssertFieldsCompatible(f0, f0_other);
+  AssertFieldsCompatible(f0, f0_with_meta1);
+  AssertFieldsCompatible(f0, f0_nn);
+  AssertFieldsCompatible(f0, f0_nt);
+  AssertFieldsCompatible(f0_nt, f0_with_meta1);
+  AssertFieldsCompatible(f0_with_meta1, f0_with_meta2);
```

Comment:
Is it desirable? We can keep this as a first approach but we may need to make it later tunable.

Reply:
Yes, in the future we can have a `struct MergeOptions`, e.g. if it should tolerate numeric promotion, etc.